Linux/sgx/test_app/enclave/tests/evp_smx.c

```diff
@@ -36,591 +36,666 @@
-            ret = -4;
-            break;
-        }
-
-        // 3. Sign
-        evp_md_ctx = EVP_MD_CTX_new();
-        if (evp_md_ctx == NULL) {
-            printf("Error: fail to create a EVP_MD_CTX\n");
-            ret = -5;
-            break;
-        }
-        if (EVP_MD_CTX_init(evp_md_ctx) != 1) {
-            printf("Error: fail to initialize the EVP_MD_CTX\n");
-            ret = -6;
-            break;	
-        }
-
-        evp_pkey_ctx = EVP_PKEY_CTX_new(evp_pkey, NULL);
-        if (evp_pkey_ctx == NULL) {
-            printf("Error: fail to create a EVP_PKEY_CTX\n");
-            ret = -7;
-            break;
-        }
-        if (EVP_PKEY_CTX_set1_id(evp_pkey_ctx, sm2_user_id, sm2_user_id_len) != 1) {
-            printf("Error: fail to set user_id to the EVP_PKEY_CTX\n");
-            ret = -8;
-            break;
-        }
-        EVP_MD_CTX_set_pkey_ctx(evp_md_ctx, evp_pkey_ctx);
-
-        if (EVP_DigestSignInit(evp_md_ctx, NULL, EVP_sm3(), NULL, evp_pkey) != 1) {
-            printf("Error: fail to initialize digest sign\n");
-            ret = -9;
-            break;			
-        }
-        if (EVP_DigestSignUpdate(evp_md_ctx, data, data_size) != 1) {
-            printf("Error: fail to update digest sign\n");
-            ret = -10;
-            break;			
-        }
-        if (EVP_DigestSignFinal(evp_md_ctx, NULL, sign_len) != 1) {
-            printf("Error: fail to finalize digest sign\n");
-            ret = -11;
-            break;			
-        }
-        if (EVP_DigestSignFinal(evp_md_ctx, signature, sign_len) != 1) {
-            printf("Error: fail to finalize digest sign\n");
-            ret = -12;
-            break;			
-        }
-
-    } while(0);
-
-    // 4. Finalize
-    BIO_free(pri_bio);
-    EC_KEY_free(ec_key);
-    EVP_PKEY_free(evp_pkey);
-    EVP_MD_CTX_free(evp_md_ctx);
-    EVP_PKEY_CTX_free(evp_pkey_ctx);
-
-    return ret;
+	int ret = 0;
+	BIO *pri_bio = NULL;
+	EVP_PKEY* evp_pkey = NULL;
+	EVP_MD_CTX *evp_md_ctx = NULL;
+	EVP_PKEY_CTX* evp_pkey_ctx = NULL;
+
+	do {
+#if OPENSSL_VERSION_NUMBER < 0x30000000
+		// 1.1 Read SM2 private key using EC_KEY
+		EC_KEY *ec_key = NULL;
+		pri_bio = BIO_new_mem_buf(private_key, -1);
+		ec_key = PEM_read_bio_ECPrivateKey(pri_bio, NULL, NULL, NULL);
+		if (ec_key == NULL) {
+			printf("Error: fail to read SM2 private key using EC_KEY\n");
+			ret = -1;
+			break;
+		}
+		// 1.2 Modify an EVP_PKEY by EC_KEY to use SM2
+		evp_pkey = EVP_PKEY_new();
+		if (evp_pkey == NULL) {
+			printf("Error: fail to create an EVP_PKEY\n");
+			ret = -2;
+			break;
+		}
+		if (EVP_PKEY_set1_EC_KEY(evp_pkey, ec_key) != 1) {
+			printf("Error: fail to set the EVP_PKEY by EC_KEY\n");
+			ret = -3;
+			break;
+		}
+		if (EVP_PKEY_set_alias_type(evp_pkey, EVP_PKEY_SM2) != 1) {
+			printf("Error: fail to modify the EVP_PKEY to use SM2\n");
+			ret = -4;
+			break;
+		}
+#elif OPENSSL_VERSION_NUMBER >= 0x30000000
```

Comment:
You can simply use only `#else` here, since it should be enough when it is `#if OPENSSL_VERSION_NUMBER < 0x30000000` above.